drivers: regulator: enable driver when compatible node is defined

jfischer-no opened the pull request:
Add regulator compatibility string to Kconfig.fixed file
and allow fixed regulator driver to be enabled when compatible
node is defined. However, remove this option from common Kconfig
as it is legitimate to enable/disable regulator support.

Change menuconfig REGULATOR_FIXED to config since this
menu nesting is not necessary.

jfischer-no committed:
drivers: regulator: enable driver when compatible node is defined

Add regulator compatibility string to Kconfig.fixed file
and allow fixed regulator driver to be enabled when compatible
node is defined. However, remove this option from common Kconfig
as it is legitimate to enable/disable regulator support.

Change menuconfig REGULATOR_FIXED to config since this
menu nesting is not necessary.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: regulator: do not explicitly enable fixed regulator driver

Do not explicitly enable fixed regulator driver for boards and tests
because it is automatically selected when corresponding node
is defined.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

pabigot commented on drivers/regulator/CMakeLists.txt:
```diff
@@ -1,6 +1,4 @@
 # Copyright 2020 Peter Bigot Consulting, LLC
 # SPDX-License-Identifier: Apache-2.0
 
-zephyr_library()
```
Every other driver subsystem puts its sources into a library.  This one should too when another type of regulator is added.  So are we just kicking the problem down the road?

Also, the next line adds `regulator_fixed.c` as a library source.  What library is it going into?

jfischer-no replied:
> Every other driver subsystem puts its sources into a library.

No, this is not consistent. I will think it over again.


> Also, the next line adds `regulator_fixed.c` as a library source. What library is it going into?

my mistake, this should probably be `zephyr_sources_ifdef(CONFIG_REGULATOR_FIXED regulator_fixed.c`) in that case.

carlescufi replied:
@tejlmand the fact that enabling `CONFIG_REGULATOR` in `tests/drivers/regulator/fixed/` makes that test fail for boards that do not have a regulator in their Devicetree is what triggers the problem with `zephyr_library()`.

tejlmand replied:
> Every other driver subsystem puts its sources into a library. 

Unfortunately not, this is very inconsistent among Zephyr driver libraries.

One root cause is probable that driver libs in general don't use depends (ask someone else why they don't as I don't know) but the sources added to the driver lib uses `zephyr_[library]_sources_ifdef()` which means that if `zephyr_library()` was used, then it could end up being empty and thus result in CMake failures.

And thus people take the easy but dirty solution of just placing it `zephyr_sources()` resulting in this mess instead.
https://github.com/zephyrproject-rtos/zephyr/issues/8825

tejlmand replied:
@jfischer-no I would prefer we keep `zephyr_library()` to avoid making more mess as described in #8825.

Looking into how this can be accomplished.

jfischer-no replied:
I agree on zephyr_library(). I reverted this changes and add comment to tests/drivers/regulator/fixed/prj.conf

jfischer-no committed:
tests: regulator: add integration platform

Add nrf52840dk_nrf52840 as integration platform since
there is already devicetree overlay for that.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tejlmand commented on tests/drivers/regulator/fixed/prj.conf:
```diff
@@ -1,5 +1,7 @@
 CONFIG_GPIO=y
 CONFIG_REGULATOR=y
+# Explicitly enable REGULATOR_FIXED as workaround for
+# "No SOURCES given to target:  drivers__regulator" CMake error.
```
please check: https://github.com/zephyrproject-rtos/zephyr/pull/34289
for a more general fix that covers the empty Zephyr lib in general for drivers.

jfischer-no replied:
rebased and removed comment

tejlmand replied:
@jfischer-no #34289 has been merged, please update the prj.conf to not contain workarounds, so we can see if there are any other outstanding issues in the build system.

pabigot reviewed (approved):
Seems fine; hope somebody takes up this driver API and gets the supporting dependencies in place so it can be used.

jfischer-no commented:
Wait until #34289 is merged.

emob-nordic commented on tests/drivers/regulator/fixed/testcase.yaml:
```diff
@@ -5,6 +5,8 @@ common:
   harness_config:
     fixture: regulator_loopback
+  integration_platforms:
+    - nrf52840dk_nrf52840
```
Not thingy52?

emob-nordic replied:
Could be both i suppose.

jfischer-no replied:
Not Thingy52. Thingy's do not have arduino gpio headers.

emob-nordic replied:
Ah yes, sorry i was thinking in terms of regualtors.
